### PR TITLE
Add qemu-io with unaligned offset testing

### PR DESCRIPTION
**Pull request by qingwangrh**
ID:2180749

Signed-off-by: qingwangrh <qinwang@redhat.com>

### CODE REVIEW

**Comment by qingwangrh** [comment]
python ConfigTest.py --testcase=block_io_on_4kdisk --guestname=RHEL.8.6.0 --clone=no --firmware=default_bios

(1/1) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.x86_64.io-github-autotest-qemu.block_io_on_4kdisk.q35: STARTED
 (1/1) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.x86_64.io-github-autotest-qemu.block_io_on_4kdisk.q35: ERROR: Command 'qemu-io -f raw -t none -c "writev -P 42 512 `cat /var/tmp/io_4kdisk`" /dev/loop0' failed.\nstdout: b'writev failed: Invalid argument\n'\nstderr: b''\nadditional_info: None (2.90 s)
RESULTS    : PASS 0 | ERROR 1 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

Failed result is expected on unaligned offset due to  product bug (https://bugzilla.redhat.com/show_bug.cgi?id=2141964) - Guest hit EXT4-fs error on host 4K disk when repeatedly hot-plug/unplug running IO disk

**Comment by qingwangrh** [comment]
@YongxueHong @XueqiangWei @zhenyzha Please help to review,thanks.

File: qemu/tests/block_io_on_4kdisk.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+"""qemu-io on 4k emulated disk"""
```
[the remark this answers is not in the document]
Comment (qingwangrh) [reply qemu/tests/block_io_on_4kdisk.py:1]: The sector-size is not a test point.

[the remark this answers is not in the document]
Comment (XueqiangWei) [reply qemu/tests/block_io_on_4kdisk.py:1]: We'd better modify the file name to make it clearer.

**Comment by qingwangrh** [comment]
Change file name to avoid confusing

python ConfigTest.py --testcase=block_io_with_unaligned_offset --guestname=RHEL.8.6.0 --firmware=default_bios --clone=no
 (1/1) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.x86_64.io-github-autotest-qemu.block_io_with_unaligned_offset.q35: STARTED
 (1/1) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.x86_64.io-github-autotest-qemu.block_io_with_unaligned_offset.q35: ERROR: Command 'qemu-io -f raw -t none -c "writev -P 42 512 `cat /var/tmp/io_4kdisk`" /dev/loop1' failed.\nstdout: b'writev failed: Invalid argument\n'\nstderr: b''\nadditional_info: None (4.21 s)
RESULTS    : PASS 0 | ERROR 1 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0


Failed result is expected on unaligned offset due to product bug (https://bugzilla.redhat.com/show_bug.cgi?id=2141964) - Guest hit EXT4-fs error on host 4K disk when repeatedly hot-plug/unplug running IO disk

File: qemu/tests/block_io_on_4kdisk.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (qingwangrh) [reply qemu/tests/block_io_on_4kdisk.py]: > We'd better modify the file name to make it clearer.

updated, thanks.

**Comment by zhenyzha** [comment]
# git am 3663.patch
```
Applying: Add qemu-io with unaligned offset testing
.git/rebase-apply/patch:53: trailing whitespace.
    pre_command += ${loop_cmd} > ${loop_dev} &&
warning: 1 line adds whitespace errors.
```

File: qemu/tests/cfg/block_io_with_unaligned_offset.cfg [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    start_vm = no
+    img_file = /var/tmp/io_4kdisk.img
+    loop_dev = /var/tmp/loop_dev
+    content_file =/var/tmp/io_4kdisk
+    io_cmd = qemu-io -f raw -t none -c "writev -P 42 %d `cat ${content_file}`" %s
+    loop_cmd = losetup -f --show --sector-size 4096 --direct-io=on ${img_file}
+    pre_command = which qemu-io && truncate -s 1G ${img_file} &&
+    pre_command += ${loop_cmd} > ${loop_dev} &&
+    pre_command += yes 512 | head -n 1024 | tr '\n' ' ' >  ${content_file}
+    post_command = losetup -d `cat  ${loop_dev}`;rm -rf ${img_file}
```
Comment (zhenyzha) [inline qemu/tests/cfg/block_io_with_unaligned_offset.cfg:12]: `cat ${loop_dev}` 

Do we want to use ' ' for this?
Other parts are ok for me.

**Comment by zhenyzha** [comment]
git am 3663.patch
Applying: Add qemu-io with unaligned offset testing

` (1/1) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.block_io_with_unaligned_offset.arm64-pci: ERROR: Command 'qemu-io -f raw -t none -c "writev -P 42 512 `cat /var/tmp/io_4kdisk`" /dev/loop1' failed.\nstdout: b'writev failed: Invalid argument\n'\nstderr: b''\nadditional_info: None (3.04 s)`

LGTM.Acked-by: Zhenyu Zhang <zhenyzha@redhat.com>

File: qemu/tests/cfg/block_io_with_unaligned_offset.cfg (hunk 2 above)
Comment (qingwangrh) [reply qemu/tests/cfg/block_io_with_unaligned_offset.cfg:12]: It is shell running symbol

File: qemu/tests/block_io_with_unaligned_offset.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+"""qemu-io with unaligned offset"""
+from avocado.utils import process
+
+
+def run(test, params, env):
+    """qemu-io with unaligned offset on simulated 4k disk.
+    Test Steps:
+    1) Create loop device with --sector-size 4096
+    2) Execute qemu-io with aligned offset
+    2) Execute qemu-io with unaligned offset
```
Comment (XueqiangWei) [inline qemu/tests/block_io_with_unaligned_offset.py:10]: \"2)\" in line 10 should be \"3)\"

Comment (qingwangrh) [reply qemu/tests/block_io_with_unaligned_offset.py]: updated, thanks.

**Comment by XueqiangWei** [comment]
LGTM,
Acked-by: xuwei@redhat.com

**Review by XueqiangWei — APPROVED** [review]

**Review by YongxueHong — APPROVED** [review]
LGTM

**Pull request merged**